src/Components/Endpoints/src/DependencyInjection/EndpointHtmlRenderer.cs

```diff
@@ -47,6 +60,37 @@ private static async Task InitializeStandardComponentServicesAsync(HttpContext h
         await componentApplicationLifetime.RestoreStateAsync(new PrerenderComponentApplicationStore());
     }
 
+    public void OnContentUpdated(Action<IEnumerable<HtmlComponentBase>> callback)
+    {
+        if (_onContentUpdatedCallback is not null)
+        {
+            // The framework is the only user of this internal API, so it's OK to have an arbitrary limit like this
+            throw new InvalidOperationException($"{nameof(OnContentUpdated)} can only be called once.");
+        }
+
+        _onContentUpdatedCallback = callback;
+    }
+
+    protected override Task UpdateDisplayAsync(in RenderBatch renderBatch)
+    {
+        var count = renderBatch.UpdatedComponents.Count;
+        if (count > 0 && _onContentUpdatedCallback is not null)
+        {
+            // TODO: Deduplicate these. There's no reason ever to include the same component more than once in a single batch,
+            // as we're rendering its whole final state, not the intermediate diffs.
+            var htmlComponents = new List<HtmlComponentBase>(count);
+            for (var i = 0; i < count; i++)
+            {
+                ref var diff = ref renderBatch.UpdatedComponents.Array[i];
+                htmlComponents.Add(new HtmlComponentBase(this, diff.ComponentId));
+            }
+
+            _onContentUpdatedCallback(htmlComponents);
+        }
+
+        return base.UpdateDisplayAsync(renderBatch);
+    }
+
```

Comment:
Since this type is internal, I would suggest you change the IEnumerable on the callback and pass in a List/Array directly. Preferably an `Array` that we pool.

Reply:
The PR auto-merged, but I'll look into this as part of the next bit of work on streaming SSR.